### PR TITLE
Add optimization rule to optimize for empty input tables

### PR DESCRIPTION
**Pull request by feilong-liu**
Add a optimizer to optimize queries with empty input.

There is already an optimization rule EliminateEmptyJoins, however it 1) only optimize for join node 2) only optimize for simple cases, for example, when the join input is an aggregation over an empty input, it will not trigger the optimization.

In this PR, I added a new rule which 1) also optimize for other nodes 2) cover more complex cases.

### Test plan - (Please fill in how you tested your changes)

Unit test and [verifier test](https://www.internalfb.com/intern/presto/verifier/results/?test_id=109889)

```
== RELEASE NOTES ==

General Changes
* Add an optimization for queries with empty input.
   The optimization is controlled by session property `simplify_plan_with_empty_input`
```

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -6562,4 +6549,24 @@ public void testSameAggregationWithAndWithoutFilter()
...
+    {
+        Session enableOptimization = Session.builder(getSession())
+                .setSystemProperty(SIMPLIFY_PLAN_WITH_EMPTY_INPUT, "true")
+                .build();
+
+        assertQuery(enableOptimization, "select o.orderkey, o.custkey, l.linenumber from orders o join (select orderkey, linenumber from lineitem where false) l on o.orderkey = l.orderkey");
+        assertQuery(enableOptimization, "select o.orderkey, o.custkey, l.linenumber from orders o left join (select orderkey, linenumber from lineitem where false) l on o.orderkey = l.orderkey");
+        assertQuery(enableOptimization, "select o.orderkey, o.custkey, l.linenumber from (select orderkey, linenumber from lineitem where false) l right join orders o on l.orderkey = o.orderkey");
+        assertQuery(enableOptimization, "select orderkey, partkey from lineitem union all select orderkey, custkey as partkey from orders where false");
+        assertQuery(enableOptimization, "select orderkey, partkey from lineitem where orderkey in (select orderkey from orders where false)");
```
Comment (kaikalur) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6564]: Also add a test for NOT IN

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 1 above)
Comment (kaikalur) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6564]: Also add for ScalarSubquery - SELECT (SELECT 1 FROM T1 WHERE false LIMIT 1) FROM T2

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+                        return convertToEmptyValuesNode(node);
+                    }
+                    else if (isEmptyNode(rewrittenLeft)) {
+                        return convertJoinToProject(node, rewrittenRight, rewrittenLeft.getOutputVariables());
+                    }
+                    else if (isEmptyNode(rewrittenRight)) {
+                        return convertJoinToProject(node, rewrittenLeft, rewrittenRight.getOutputVariables());
+                    }
+                    break;
+                default:
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java:180]: Cross join is Inner join with empty EquiJoinClause in our representation, hence will be covered too. Also added a cross join in the query plan test.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 1 above)
Comment (feilong-liu) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6564]: Added

**Review by kaikalur — APPROVED** [review]
nice!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            return convertToEmptyNodeIfInputEmpty(node, context);
+        }
+
+        @Override
+        public PlanNode visitLimit(LimitNode node, RewriteContext<Void> context)
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java:268]: Btw we can always remove Limit from Limit(Values()) by dropping extra rows. We don't have to implement that in this PR(or this rule) however

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSimplifyPlanWithEmptyInput.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    @Test
+    public void testInnerJoinWithOverEmptyAggregation()
+    {
+        assertPlan("select o.orderkey, o.custkey, l.linenumber from orders o join (select orderkey, max(linenumber) as linenumber from lineitem where false group by orderkey) l on o.orderkey = l.orderkey",
+                enableOptimization(),
+                output(
+                        ImmutableList.of("orderkey", "custkey", "linenumber"),
+                        values("orderkey", "custkey", "linenumber")));
+    }
+}
```
Comment (pranjalssh) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSimplifyPlanWithEmptyInput.java:193]: Also add some unit tests for simple nodes like limit/sort/filter/window. You can have it all in one sql query, where it all boils down to just empty values().

Comment (feilong-liu) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSimplifyPlanWithEmptyInput.java:216]: Added

**Review by kaikalur — APPROVED** [review]

**Review by pranjalssh — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java (hunk 3 above)
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java:268]: Yeah, I meant something else. Limit 10 -> Values(100 rows) -> Values(10) rows

But that is not related to this PR, and probably not that useful as well

**Comment by rohanpednekar** [comment]
@feilong-liu again what is the default value for the property? We need to update this in release notes.

**Pull request merged**